python/mlc_chat/compiler/quantization/ft_quantization.py

```diff
@@ -0,0 +1,399 @@
+        self.q_weight = nn.Parameter(
+            (in_features, tir.ceildiv(out_features, config.num_elem_per_storage)),
+            config.storage_dtype,
+        )
+        self.q_scale = nn.Parameter(
+            (tir.ceildiv(in_features, cur_group_size), out_features), config.model_dtype
+        )
+        if bias:
+            self.bias = nn.Parameter(
+                (out_features,), config.model_dtype if out_dtype is None else out_dtype
+            )
+        else:
+            self.bias = None
+
+    @staticmethod
+    def from_linear(src: nn.Linear, config: FTQuantize) -> "FTQuantizeLinear":
+        """
+        Converts a non-quantized nn.Linear to a group quantized GroupQuantizeLinear
+
+        Parameters
+        ----------
+        src : nn.Linear
+            The non-quantized nn.Linear.
+
+        config : GroupQuantize
+            The group quantization config.
+
+        Returns
+        -------
+        ret : GroupQuantizeLinear
+            The group quantized GroupQuantizeLinear layer.
+        """
+        quantized_linear = FTQuantizeLinear(
+            in_features=src.in_features,
+            out_features=src.out_features,
+            config=config,
+            bias=getattr(src, "bias", None) is not None,
+            out_dtype=src.out_dtype,
+        )
+        if quantized_linear.bias is not None:
+            quantized_linear.bias.attrs = src.bias.attrs
+        # if "shard_strategy" in src.weight.attrs:
+        #     shard = src.weight.attrs["shard_strategy"]
+        #     _apply_sharding(shard, f"{shard.name}_q_weight", quantized_linear.q_weight)
+        #     _apply_sharding(shard, f"{shard.name}_q_scale", quantized_linear.q_scale)
+        return quantized_linear
+
+    def forward(self, x: nn.Tensor) -> nn.Tensor:  # pylint: disable=invalid-name
+        """
+        Forward method for group quantized linear layer.
+
+        Parameters
+        ----------
+        x : nn.Tensor
+            The input tensor.
+
+        Returns
+        -------
+        ret : nn.Tensor
+            The output tensor for the group quantized linear layer.
+        """
+        return extern_op.decode_matmul(
+            x, self.q_weight, self.q_scale, self.bias, self.config.group_size
+        )
+
+    def to(self, dtype: Optional[str] = None) -> None:
+        """
+        Override to() such that we do not convert bias if there is an out_dtype.
+        Otherwise, we might run into dtype mismatch when computing x + self.bias.
+        """
+        self.q_weight.to(dtype=dtype)
+        self.q_scale.to(dtype=dtype)
+        if self.bias is not None and self.out_dtype is None:
+            self.bias.to(dtype=dtype)
+        if dtype is not None and isinstance(getattr(self, "dtype", None), str):
+            self.dtype = dtype  # pylint: disable=attribute-defined-outside-init
+
+
+def _apply_sharding(shard, name: str, weight: nn.Parameter):
```

Comment:
did we validate if it works? my impression is that `KN32k` layout may need more careful handling